File: 2022-2023/sourcecode/lab/lab11/VeryNobleAppJava/app/src/main/java/pl/udu/uwr/pum/verynobleappjava/ui/fragments/nobelaward/NobelAwardFragment.java
package pl.udu.uwr.pum.verynobleappjava.ui.fragments.nobelaward;

import android.os.Bundle;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;
import androidx.lifecycle.ViewModelProvider;
import androidx.recyclerview.widget.LinearLayoutManager;

import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import pl.udu.uwr.pum.verynobleappjava.R;
import pl.udu.uwr.pum.verynobleappjava.adapters.nobelprize.NobelPrizeAdapter;
import pl.udu.uwr.pum.verynobleappjava.adapters.nobelprize.NobelPrizeComparator;
import pl.udu.uwr.pum.verynobleappjava.adapters.nobelprizelaureates.LaureateAdapter;
import pl.udu.uwr.pum.verynobleappjava.adapters.nobelprizelaureates.LaureateComparator;
import pl.udu.uwr.pum.verynobleappjava.data.NobelAwardsResponse;
import pl.udu.uwr.pum.verynobleappjava.databinding.FragmentNobelAwardBinding;
import pl.udu.uwr.pum.verynobleappjava.databinding.FragmentNobleAwardsFragmentBinding;
import pl.udu.uwr.pum.verynobleappjava.ui.fragments.nobelawards.NobelAwardsViewModel;
import pl.udu.uwr.pum.verynobleappjava.util.Constants;

public class NobelAwardFragment extends Fragment {

    private FragmentNobelAwardBinding binding;

    private String category;
    private String awardYear;

    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        binding = FragmentNobelAwardBinding.inflate(inflater, container, false);
        category = requireArguments().getString("category");
        awardYear = requireArguments().getString("awardYear");
        return binding.getRoot();
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        NobelAwardViewModel viewModel = new ViewModelProvider(this).get(NobelAwardViewModel.class);

        viewModel.getNobelPrize(Integer.parseInt(awardYear), Constants.categories.get(Constants.CATEGORIES.valueOf(category.toUpperCase().replaceAll("\\s+",""))));

        LaureateAdapter adapter = new LaureateAdapter(new LaureateComparator());

        setupRecyclerView(adapter);

        viewModel.getNobelPrize().observe(getViewLifecycleOwner(), nobelAwardsResponse -> {
            NobelAwardsResponse.NobelPrizes item = nobelAwardsResponse.getNobelPrizes().stream().findFirst().orElse(null);
            if (item != null) {
                if (item.getCategoryFullName() != null && item.getCategoryFullName().getEn() != null)
                    binding.categoryFullNameTextView.setText(item.getCategoryFullName().getEn());
                binding.dateAwardTextView.setText(item.getDateAwarded());
                adapter.submitList(item.getLaureates());
            }
        });

    }

    private void setupRecyclerView(LaureateAdapter adapter){
        binding.nobelAwardRV.setAdapter(adapter);
        binding.nobelAwardRV.setLayoutManager(new LinearLayoutManager(requireContext()));
    }
}

File: 2022-2023/sourcecode/lab/lab11/VeryNobleAppKotlin/app/src/main/java/pl/udu/uwr/pum/verynobleappkotlin/ui/fragments/nobelaward/NobelAwardFragment.kt
package pl.udu.uwr.pum.verynobleappkotlin.ui.fragments.nobelaward

import android.os.Bundle
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.fragment.app.Fragment
import androidx.fragment.app.viewModels
import androidx.recyclerview.widget.LinearLayoutManager
import pl.udu.uwr.pum.verynobleappkotlin.adapters.nobleprizelaureates.LaureateAdapter
import pl.udu.uwr.pum.verynobleappkotlin.adapters.nobleprizelaureates.LaureateComparator
import pl.udu.uwr.pum.verynobleappkotlin.databinding.FragmentNobelAwardBinding
import pl.udu.uwr.pum.verynobleappkotlin.util.Cat
import pl.udu.uwr.pum.verynobleappkotlin.util.Resource
import pl.udu.uwr.pum.verynobleappkotlin.util.categories

class NobelAwardFragment : Fragment() {

    private lateinit var binding: FragmentNobelAwardBinding

    private val nobelPrizeViewModel: NobelPrizeViewModel by viewModels()
    private val TAG = "NobelAwardFragment"

    private val category: String? by lazy { requireArguments().getString("category") }
    private val awardYear: String? by lazy { requireArguments().getString("awardYear") }

    override fun onCreateView(
        inflater: LayoutInflater, container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View {
        binding = FragmentNobelAwardBinding.inflate(inflater, container, false)
        return binding.root
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        nobelPrizeViewModel.getNobelPrize(year= awardYear?.toInt()!!, category= categories[Cat.valueOf(category!!.uppercase().filterNot { it.isWhitespace() })]!!)

        val laureateAdapter = LaureateAdapter(LaureateComparator())
        setupRecyclerView(laureateAdapter)

        observeNoblePrize(laureateAdapter)
    }

    private fun hideProgressBar(){
        binding.nobelPrizeProgressBar.visibility = View.INVISIBLE
    }

    private fun showProgressBar(){
        binding.nobelPrizeProgressBar.visibility = View.VISIBLE
    }

    private fun observeNoblePrize(laureateAdapter: LaureateAdapter) {
        nobelPrizeViewModel.noblePrize.observe(viewLifecycleOwner) { response ->
            when (response) {
                is Resource.Success -> {
                    hideProgressBar()
                    response.data?.let {
                        binding.categoryFullNameTextView.text = it.nobelPrizes.first().categoryFullName.en
                        binding.dateAwardTextView.text = it.nobelPrizes.first().dateAwarded
                        laureateAdapter.submitList(it.nobelPrizes.first().laureates)
                    }
                }
                is Resource.Error -> {
                    hideProgressBar()
                    response.message?.let { Log.e(TAG, "Error occurred: $it") }
                }
                is Resource.Loading -> showProgressBar()
            }
        }
    }

    private fun setupRecyclerView(laureateAdapter: LaureateAdapter) {
        binding.nobelAwardRV.apply {
            adapter = laureateAdapter
            layoutManager = LinearLayoutManager(requireContext())
        }
    }
}